arangod/RestServer/LanguageCheckFeature.cpp

```diff
@@ -56,70 +59,127 @@ arangodb::Result readLanguage(arangodb::ArangodServer& server,
 }
 
 /// @brief writes the default language to file
 ErrorCode writeLanguage(arangodb::ArangodServer& server,
-                        std::string const& language) {
+                        std::string const& lang, bool isDefaultLang) {
   auto& databasePath = server.getFeature<arangodb::DatabasePathFeature>();
   std::string filename = databasePath.subdirectoryName("LANGUAGE");
 
   // generate json
   VPackBuilder builder;
   try {
     builder.openObject();
-    builder.add("default", VPackValue(language));
+    if (isDefaultLang) {
+      builder.add("default", VPackValue(lang));
+    } else {
+      builder.add("icu-language", VPackValue(lang));
+    }
+
     builder.close();
   } catch (...) {
     // out of memory
-    LOG_TOPIC("4fa50", ERR, arangodb::Logger::CONFIG)
-        << "cannot save default language in file '" << filename
-        << "': out of memory";
+    if (isDefaultLang) {
+      LOG_TOPIC("4fa50", ERR, arangodb::Logger::CONFIG)
+          << "cannot save default language in file '" << filename
+          << "': out of memory";
+    } else {
+      LOG_TOPIC("4fa51", ERR, arangodb::Logger::CONFIG)
+          << "cannot save icu language in file '" << filename
+          << "': out of memory";
+    }
+
     return TRI_ERROR_OUT_OF_MEMORY;
   }
 
   // save json info to file
   LOG_TOPIC("08f3c", DEBUG, arangodb::Logger::CONFIG)
-      << "Writing default language to file '" << filename << "'";
+      << "Writing language to file '" << filename << "'";
   bool ok = arangodb::basics::VelocyPackHelper::velocyPackToFile(
       filename, builder.slice(), true);
   if (!ok) {
     LOG_TOPIC("c2fd7", ERR, arangodb::Logger::CONFIG)
-        << "could not save default language in file '" << filename
+        << "could not save language in file '" << filename
         << "': " << TRI_last_error();
     return TRI_ERROR_INTERNAL;
   }
 
   return TRI_ERROR_NO_ERROR;
 }
 
-std::string getOrSetPreviousLanguage(arangodb::ArangodServer& server,
-                                     std::string const& input) {
-  std::string language;
-  arangodb::Result res = ::readLanguage(server, language);
+std::tuple<std::string, bool> getOrSetPreviousLanguage(
+    arangodb::ArangodServer& server, std::string const& collatorLang,
+    bool isDefaultLangSet, bool isIcuLangSet) {
+  std::string readDefaultLanguage = {};
+  std::string readIcuLanguage = {};
+  arangodb::Result res =
+      ::readLanguage(server, readDefaultLanguage, readIcuLanguage);
+
+  auto error = [] {
+    LOG_TOPIC("55a69", FATAL, arangodb::Logger::CONFIG)
+        << "Current language option is differ from option at initial launch";
+    FATAL_ERROR_EXIT();
+  };
+
   if (res.ok()) {
-    return language;
+    if (!readDefaultLanguage.empty()) {
+      // if default language is not set, we will treat it as default
+      // but if icu language is specified, this is an error
+      if (!isIcuLangSet) {
+        return std::make_tuple(readDefaultLanguage, true);
```

Comment:
Don't use make function expect shared_ptr plz, and maybe unique_ptr
It's unnecessary and slow for compilation, just use uniform initialization `{}`